feat: add `upsertUnique` and `loadUnique` APIs to CoMap (#2493)
garden-co/jazz: homepage/homepage/content/docs/using-covalues/comaps.mdx

console.log(formatProjectDuration(project, "short")); // "3 days"
```
</CodeGroup>

#### Upserting

> Consider a scenario where one needs to update a CoMap if it exists, or create it if it doesn't:

Reviewer: The docs should focus more on the real value of the API, which is not upserting, but make it possible to refer a coValue with human-firendly strings.

- Example from learnAnything on how to do slugs [link](https://github.com/learn-anything/attempt/blob/e6ab7a8aa44cfc77c50b1e3c7e5bbd17dd7b8905/web/app/routes/_layout/_pages/(topic)/%24.tsx#L22)
- Example from our cursor example on how to make a "global" covalue [link](https://github.com/garden-co/jazz/blob/main/examples/multi-cursors/src/utils/loadCursorContainer.ts#L49)

The second one should be ported to upsertUnique

console.log(formatProjectDuration(project, "short")); // "3 days"
```
</CodeGroup>

#### Uniqueness

> Identifying CoMaps by their CoValue IDs can be inconvenient if using an externally-generated identifier is preferrable, such as when:

Reviewer 2: ```suggestion
CoMaps are typically created with a CoValue ID that acts as an opaque UUID, by which you can then load them. However, there are situations where it is preferable to load CoMaps using a custom identifier:
```

The sentence is a bit convoluted, here's a simplification suggestion

- The CoMaps have user-generated identifiers, such as a slug
- The CoMaps have identifiers referring to equivalent data in an external system
- The CoMaps have human-readable & application-specific identifiers

>   - If an application has CoValues used by every user, referring to it by a unique name (eg, `"my-global-comap"`) can be more convenient than using a CoValue ID

Reviewer 2: ```suggestion
  - If an application has CoValues used by every user, referring to it by a unique *well-known* name (eg, `"my-global-comap"`) can be more convenient than using a CoValue ID
```

"well-known" is a typical way to refer to this kind of abstraction

Consider a scenario where one wants to identify a CoMap using some unique identifier that isn't the Jazz CoValue ID:

<CodeGroup>
```ts twoslash
import { co, z, Group, ID } from "jazz-tools";
const Task = co.map({
  text: z.string(),
});
// ---cut---
// This will not work as `learning-jazz` is not a CoValue ID
const myTask = await Task.load("learning-jazz");
```
</CodeGroup>

To make it possible to use human-readable identifiers Jazz lets you to define a `unique` property on CoMaps.

> Then the CoValue id will be generated based on the `unique` property and the owner of the CoMap.

Reviewer 2: ```suggestion
Then the CoValue ID is deterministically derived from the `unique` property and the owner of the CoMap.
```

<CodeGroup>
```ts twoslash
import { co, z, Group, ID } from "jazz-tools";
const Task = co.map({
  text: z.string(),
});
const Project = co.map({
  name: z.string(),
  startDate: z.date(),
  endDate: z.optional(z.date()),
});
const project = Project.create({
  name: "My project",
  startDate: new Date("2025-04-01"),
  endDate: new Date("2025-04-04"),
});
// ---cut---
const learnJazzTask = await Task.loadUnique(
  "learning-jazz",
  project._owner.id
);
```
</CodeGroup>

It's also possible to combine the create+load operation using `CoMap.upsertUnique`:

<CodeGroup>
```ts twoslash
import { co, z, Group, ID } from "jazz-tools";
const Task = co.map({
  text: z.string(),
});
const Project = co.map({
  name: z.string(),
  startDate: z.date(),
  endDate: z.optional(z.date()),
});
const project = Project.create({
  name: "My project",
  startDate: new Date("2025-04-01"),
  endDate: new Date("2025-04-04"),
});
// ---cut---
const learnJazzTask = await Task.upsertUnique(
  {
    value: {
      text: "Let's learn some Jazz!",
    },
    unique: "learning-jazz",
    owner: project._owner,
  }
);
```

> </CodeGroup>

Reviewer 2: ```suggestion
</CodeGroup>

**Caveat:** The `unique` parameter acts as an *immutable* identifier - i.e. the same `unique` parameter in the same `Group` will always refer to the same CoValue.

To make dynamic renaming possible, you can create an indirection where a stable CoMap identified by a specific value of `unique` is simply a pointer to another CoMap with a normal, dynamic CoValue ID. This pointer can then be updated as desired by users with the corresponding permissions.
```

This is kind of a logical implication, but worth spelling it out
